### PR TITLE
Add support for SVG images

### PR DESCRIPTION
**Pull request by hopsoft**

### CODE REVIEW

File: Gemfile [hunk 1]
```diff
@@ -6,10 +6,11 @@ git_source(:github) { |repo| "https://github.com/#{repo}.git" }
 ruby "2.6.4"
 
 gem "active_link_to", "~> 1.0.5"
+gem "active_storage_svg_sanitizer", "~> 0.1.0"
 gem "acts_as_commentable_with_threading", "~> 2.0.1"
 gem "addressable", "2.6.0"
 gem "airbrake", "~> 9.3"
-gem "aws-sdk-s3", "~> 1.46.0", require: false
+gem "aws-sdk-s3", "~> 1.48.0", require: false
```
Comment (dahal) [inline Gemfile]: Does it require gem update? If it's not part of it, why not make a separate PR for it?

File: app/analyzers/creative_image_analyzer.rb [hunk 2]
```diff
@@ -1,14 +1,20 @@
...
+    super.dup.tap do |md|
+      md[:name] = @blob.filename
 
-    md[:format] = "icon" if md[:width].to_i == 20 && md[:height] == 20
-    md[:format] = "icon" if md[:width].to_i == 40 && md[:height] == 40
-    md[:format] = "small" if md[:width].to_i == 200 && md[:height] == 200
-    md[:format] = "large" if md[:width].to_i == 260 && md[:height] == 200
-    md[:format] = "wide" if md[:width].to_i == 512 && md[:height] == 320
-    md
+      if md[:height].present? && md[:width].present?
```
Comment (dahal) [inline app/analyzers/creative_image_analyzer.rb]: You can just use `md.try(:[], :height).present?` for short.

* https://everydayrails.com/2011/04/28/rails-try-method.html

File: Gemfile (hunk 1 above)
Comment (hopsoft) [reply Gemfile]: Mostly because we're an extremely small shop and we're also lazy.

File: app/analyzers/creative_image_analyzer.rb (hunk 2 above)
Comment (hopsoft) [reply app/analyzers/creative_image_analyzer.rb]: That's a cool trick but I don't think it makes it any more readable.

Comment (dahal) [reply app/analyzers/creative_image_analyzer.rb]: Right, it prolly was late at night when i reviewed this, for some reason I thought you were just checking for `height`, just noticed you are checking for `weight` presence as well.

File: app/jobs/create_sponsor_svg_job.rb [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+    return if blob.metadata[:format] == ENUMS::IMAGE_FORMATS::SPONSOR
+
+    users = blob.attachments.where(record_type: "User").map(&:record)
+    sponsor_svg = SvgsController.render(partial: "/svgs/sponsors/default", locals: {logo: embedded_svg(blob).html_safe})
+    filename = "sponsor-#{blob.filename.base}.svg"
+    Tempfile.open(filename, Rails.root.join("tmp")) do |file|
+      file.write sponsor_svg
+      users.each do |user|
+        file.rewind
+        next unless AuthorizedUser.new(user).can_create_sponsor_svg?
```
Comment (hopsoft) [inline app/jobs/create_sponsor_svg_job.rb]: Currently the auth check fails for non admins. I wonder if we should simply let this run for everyone. Thoughts?

Comment (coderberry) [reply app/jobs/create_sponsor_svg_job.rb]: Probably.. although we need to make sure that they can't swap out the creative mid-campaign. It still must be approved.

File: app/models/concerns/authorizers/image.rb [hunk 4]
```diff
@@ -9,5 +9,9 @@ def can_destroy_image?(image)
       return true if can_admin_system?
       image.record == user
     end
+
+    def can_create_sponsor_svg?
```
Comment (coderberry) [inline app/models/concerns/authorizers/image.rb]: We may want to allow this. We just need to have the creative status as "active". As long as it remains pending and locked when active, I think we can open this up.

File: app/models/organization_report.rb [hunk 5]
```diff
@@ -48,9 +49,6 @@ def campaigns
   end
 
   # protected instance methods ................................................
-  protected
```
Comment (coderberry) [inline app/models/organization_report.rb]: Sorry :(

File: app/models/property.rb [hunk 6]
```diff
@@ -138,6 +139,14 @@ class << self
 
   # public instance methods ...................................................
 
+  def host
+    URI.parse(url).host
+  end
+
+  def restrict_to_sponsor_campaigns?
+    host == "github.com"
```
Comment (coderberry) [inline app/models/property.rb]: we probably do not want this. The GitHub Readme is used on an additional 180+ sites. We want it to work on all. We also will allow the sponsor placement on GitLab, etc.

File: config/routes.rb [hunk 7]
```diff
@@ -107,6 +107,8 @@
     resource :impression_uplifts, only: [:create], path: "/uplift"
   end
 
+  get "/impressions/click"
```
Comment (coderberry) [inline config/routes.rb]: Is this intentional?

File: test/assets/images/logo-digitalocean.svg [hunk 8]
```diff
@@ -0,0 +1 @@
...13.9 0 0 1 9 3 9 9 0 0 1 3.49 7.47v3.49a38.13 38.13 0 0 0-12.45-2q-12.7 0-20.92 6c-5.48 4-8 10-8 16.94 0 6.48 2 11.46 6.47 15.44s10 5.48 16.43 5.48 12.45-2.49 18.43-7v5.48h16.93V91.79c0-8-2.49-14.45-8-18.93m-30.38 36.86c2-1.49 4.48-2 8-2a42.37 42.37 0 0 1 12.95 2.49v6.48a22 22 0 0 1-14.94 5.48c-3 0-5.48-.5-7-2s-2.49-3-2.49-5a8.17 8.17 0 0 1 3.49-5.48m104.54-35.84c-5-5.48-11.45-8-19.92-8q-10.46 0-16.43 6v-4h-16.94v68.25h17.43V98.26c0-5 1-9.47 3.49-12.45a12.54 12.54 0 0 1 10-4.48 10.94 10.94 0 0 1 9 4 18 18 0 0 1 3.49 11v39.85H1002V96.27q-.75-14.2-7.47-22.42m-423.31-1c-5-4.48-12.45-6.48-21.41-6.48a43.59 43.59 0 0 0-15.94 3.49c-4.48 2.49-9 6-11.45 11l11 10.46c4.48-7.47 9.46-10 16.43-10a13.9 13.9 0 0 1 9 3 9 9 0 0 1 3.49 7.47v3.49a38.13 38.13 0 0 0-12.45-2q-12.7 0-20.92 6c-5.48 4-8 10-8 16.94 0 6.48 2 11.46 6.47 15.44s10 5.48 16.43 5.48 12.45-2.49 18.43-7v5.48h16.93V91.79c0-8-3-14.45-8-18.93m-30.88 36.86c2-1.49 4.48-2 8-2a42.37 42.37 0 0 1 12.95 2.49v6.48a22 22 0 0 1-14.94 5.48c-3 0-5.48-.5-7-2s-2.49-3-2.49-5a14.84 14.84 0 0 1 3.49-5.48m132 27.9a49.82 49.82 0 1 1 49.8-49.82 50 50 0 0 1-49.8 49.82m0-82.69a32.38 32.38 0 1 0 32.33 32.4 32.31 32.31 0 0 0-32.37-32.38" fill="#0080ff"/></svg>
```
Comment (coderberry) [inline test/assets/images/logo-digitalocean.svg]: Should we keep all these in the code base?

File: app/models/property.rb (hunk 6 above)
Comment (hopsoft) [reply app/models/property.rb]: The sponsor will show wherever the readme is hosted. 

This logic is simply checking the canonical property domain. For example, consider a project who's canonical repo is at GitHub, but is also mirrored on Gitlab... and it's an NPM package. The property/project would be listed in CodeFund with the canonical GitHub URL, but the readme (including sponsor) will be rendered on GitHub, Gitlab, NPM, and any other sites that render the readme etc... 

Note that we can expand the definition of canonical sources to include Gitlab etc... as we find projects that want to list their primary URL as being from those domains. For example:

```ruby
 def restrict_to_sponsor_campaigns?
  %w[github.com gitlab.com bitbucket.org].include? host
end
```

File: test/assets/images/logo-digitalocean.svg (hunk 8 above)
Comment (hopsoft) [reply test/assets/images/logo-digitalocean.svg]: I've got them in there for testing purposes. Unfortunately, I removed those tests for now, but these images don't need to be brand specific.

File: app/controllers/advertisement_clicks_controller.rb [hunk 9]
```diff
@@ -21,30 +21,58 @@ def show
...
+    request.path.include? "/visit-sponsor"
+  end
+
   def set_variables
-    @campaign = Campaign.select(:id, :name, :url).find_by(id: params[:campaign_id])
-    @creative = Creative.select(:id, :name).find_by(id: params[:creative_id])
     @property = Property.select(:id, :name, :url).find_by(id: params[:property_id])
+
+    if sponsor?
+      @campaign = @property.current_sponsor_campaign
```
Comment (coderberry) [inline app/controllers/advertisement_clicks_controller.rb]: If it's a fallback sponsor, do they link to our homepage?

File: test/controllers/advertisement_clicks_controller_test.rb [hunk 10]
```diff
@@ -145,4 +146,50 @@ class AdvertisementClicksControllerTest < ActionDispatch::IntegrationTest
...
+  test "sponsor advertisement click without active campaign" do
+    property = amend(properties: :website, url: "https://github.com/gitcoinco/code_fund_ads")
+    ip = ip_address("US")
+
+    Impression.delete_all
+    assert Impression.count == 0
+    assert property.restrict_to_sponsor_campaigns?
+    get sponsor_visit_url(property), headers: {"REMOTE_ADDR": ip, "User-Agent": "Rails/Minitest"}
+    assert_response :found
+    assert headers["Location"] == advertisers_url
```
Comment (coderberry) [inline test/controllers/advertisement_clicks_controller_test.rb]: This answers my earlier question

File: app/controllers/advertisement_clicks_controller.rb (hunk 9 above)
Comment (hopsoft) [reply app/controllers/advertisement_clicks_controller.rb]: They get sent to the advertiser page if we hit the catch all.
